README.md
# Voice Actor Chatter for FVTT

![Latest Release Download Count](https://img.shields.io/github/downloads/p4535992/foundryvtt-voice-actor-chatter/latest/module.zip?color=2b82fc&label=DOWNLOADS&style=for-the-badge) 

[![Forge Installs](https://img.shields.io/badge/dynamic/json?label=Forge%20Installs&query=package.installs&suffix=%25&url=https%3A%2F%2Fforge-vtt.com%2Fapi%2Fbazaar%2Fpackage%2Ffoundryvtt-voice-actor-chatter&colorB=006400&style=for-the-badge)](https://forge-vtt.com/bazaar#package=foundryvtt-voice-actor-chatter) 

![Foundry Core Compatible Version](https://img.shields.io/badge/dynamic/json.svg?url=https%3A%2F%2Fraw.githubusercontent.com%2Fp4535992%2Ffoundryvtt-voice-actor-chatter%2Fmaster%2Fsrc%2Fmodule.json&label=Foundry%20Version&query=$.compatibleCoreVersion&colorB=orange&style=for-the-badge)

![Latest Version](https://img.shields.io/badge/dynamic/json.svg?url=https%3A%2F%2Fraw.githubusercontent.com%2Fp4535992%2Ffoundryvtt-voice-actor-chatter%2Fmaster%2Fsrc%2Fmodule.json&label=Latest%20Release&prefix=v&query=$.version&colorB=red&style=for-the-badge)

[![Foundry Hub Endorsements](https://img.shields.io/endpoint?logoColor=white&url=https%3A%2F%2Fwww.foundryvtt-hub.com%2Fwp-json%2Fhubapi%2Fv1%2Fpackage%2Ffoundryvtt-voice-actor-chatter%2Fshield%2Fendorsements&style=for-the-badge)](https://www.foundryvtt-hub.com/package/foundryvtt-voice-actor-chatter/)

This module was born as a union of the features of the two [Voice Actor for FVTT by Blitz](https://github.com/BlitzKraig/fvtt-VoiceActor) and [NPC Chatter](https://github.com/cswendrowski/FoundryVtt-Npc-Chatter) modules.

This module arises from the need together with the [Trigger Happy module](https://github.com/League-of-Foundry-Developers/fvtt-module-trigger-happy) to randomize a set of voices for the same actor through a customizable rollTable. Basicaly i want when click on a actor to have a random prepared audio to run.

This work even with external audio just set the source on the roll table.

Here some macro for give the idea

```
var token = canvas.tokens.placeables.filter(x => x.data.name == "MerchantWeapon01")[0];
game.voiceActorChatter.tokenChatter(token);
```

## Installation

It's always easiest to install modules from the in game add-on browser.

To install this module manually:
1.  Inside the Foundry "Configuration and Setup" screen, click "Add-on Modules"
2.  Click "Install Module"
3.  In the "Manifest URL" field, paste the following url:
`https://raw.githubusercontent.com/p4535992/foundryvtt-voice-actor-chatter/master/src/module.json`
4.  Click 'Install' and wait for installation to complete
5.  Don't forget to enable the module in game using the "Manage Module" button

# Important note for servers accessed over HTTP/IP

Chrome and Firefox will not ask you to grant microphone permissions if you do not have an SSL enabled site, they will just refuse automatically, breaking the module.

If you only want the GM to record via the module, use `localhost` to access Foundry if possible.

If this is not possible due to external hosting etc., I have discovered a workaround for Chrome, and a slightly worse workaround for Firefox.

## *PROCEED AT YOUR OWN RISK*

These workarounds should be safe, but I am not a security expert.
I would advise running via localhost, or setting up SSL for your server.

## Chrome (Secure-ish workaround)

* Navigate to `chrome://flags/#unsafely-treat-insecure-origin-as-secure`
* Enable the flag and add your foundry IP and port: `http://ip.address.here:30000`
* Relaunch, and Chrome will now ask for mic permissions when you try to record to an actor

## Firefox (Less secure workaround)

* Navigate to `about:config`
* Search for `insecure`
* Set `media.devices.insecure.enabled` to `true`
* Set `media.getusermedia.insecure.enabled` to `true`
* Refresh foundry page, Firefox will now ask for mic permission when you try to record to an actor

## The difference

Chrome lets you whitelist the server directly, meaning the security holes this potentially exposes can't be taken advantage of by other insecure sites.
Firefox only lets you blanket allow insecure `getusermedia` and `devices`, meaning an insecure site could theoretically try to take advantage of this.

Because they are both permissions-based, you need to actually allow the site to use them, so the risk is pretty low. HOWEVER, **you do this at your own risk!**
I would advise looking into setting up an HTTPS foundry server.

# Now that's out of the way, on to the module!

## What's it for?

This module is my very late and unofficial D20 day contribution!
I wrote it in a couple of hours, hoping it would help me dip my feet back into Foundry module development.
It's pretty simple, a little hacky, and contains some less than stellar coding practices... BUT, it's also pretty useful!

In a nutshell, I find myself constantly asking my players "What did I make this guy sound like again?..." whenever we revisit an NPC.
This module should help with that problem, allowing the GM to record a sample of an NPC's voice directly in their sheet, or in a journal entry, and play it back privately whenever they want.

## What does it do?

Adds two new buttons to the titlebar of an actor sheet and journal entry, allowing you to record and play-back a clip.

Records directly from whatever your browsers input source is, and saves the file to a VoiceActor directory in your userdata root.

Clips can be 30 seconds max, and can be ended early by clicking the stop button.

Linked actors will have a single recorded clip, saved using their actor ID.

Journal entries will have a single recorded clip, saved using their entry ID in a Journal/ directory inside the VoiceActor/ directory

Unlinked actors will have a clip saved based on their ID and actor name.
This means unlinked actors with the same name will share the same clip, but changing an actors name in their unlinked sheet after spawning will allow you to record a clip for that actor specifically (and any other actors of that 'type' with the same name)

e.g. If you change one unlinked Goblin's name to "Boblin" in his actor sheet, he can have his own special clip recorded. All of the goblins named "Goblin" will share a single clip.

## How do I use it?

- Install & enable the module
- Make sure to create a folder **'Voice Actor'** on the rolltable sidebar
- Double click a token to open their actor sheet (or open a journal entry)
- Note the 2 new buttons at the top left of the window
- Click the microphone button to record (up to 10 seconds)
- Click again to stop recording
- Click play to playback the recorded audio (does not broadcast, only plays for the GM)
- Shift-click play to broadcast to all players (note that this isn't what the module is designed for, but it was easy enough to implement, so I've jammed it in there in case you want to give your NPCs a repeatable greeting, or give your ogres a nice roar)
- Shift-click the microphone button to record a new clip, overwriting the previous clip for that actor

## What else should I know?

Linked actor tokens will save their clip using the actor ID. This means your BBEG NPC can have a clip recorded, and will play back that same clip on every spawned token.

Unlinked actor tokens will save their clip using the actor ID AND the actor name. This means, for example, all of your "Commoner"s will share a clip, except tokens you spawn and then change the name of. Maybe there's a memorable NPC that uses the Commoner statblock - no problem! Spawn a Commoner, double click the token, change the token name to your NPC's name, then record your clip!

# Macro Voice Actor Chatter

# Usage

Chatter can only be triggered via scripting. There are a few Macros included as examples - most work out of the box, but some require additional setup.

The easiest macro is the "Timed Global Chatter" macro - just slap it when a Scene Loads, and they will start chattering!

## Trigger Happy

[Trigger Happy](https://github.com/kakaroto/fvtt-module-trigger-happy) can trigger Voice Actor Chatter as well. Here's an example to get you started:

When an Actor walks into a Room (defined by an invisble actor), have a specific Token chatter: 

```
@Actor[TriggerA] @Macro[OXyjmVhEGo3eTaJz]{Specific Token Chatter}
```

# Features

- Allow Owner recording : Enable to let players record and playback on their owned actors. Requires FILES_UPLOAD permission for players.

- Allow Limited & Observer playback: Enable to let players playback clips on actors they have Limited or Observer permissions on. Players cannot record to these actors.

- Custom directory path: Custom directory path

- Disable header Sheet Buttons: Disable header Sheet Buttons after you are ok with the chatter audio file

- Upload file util: A setting for upload file audio on the server

- Integration with Polyglot: Set up different rolltable one for each language and change the audio file with customized one for that language.

- Integration with NPC Chatter: Integration with NPC Chatter module

# API

## Options Parameters

| Key    | SubKey  | Type              | Description                                                     | Default |
|:------:|:-------:|:-----------------:|:---------------------------------------------------------------:|:-------:|
| toAll  |         | boolean           | Broadcast audio to all players                                  | false   |
| langs  |         | array (of String) | Explicit languages known to target ignore the polyglot settings | []      |

## Actor Voice Global Chatter

Picks a random Chatter Table belonging to a random Actor on the first active scene and play a audio file with rolled Text off of the random Chatter Table.

```js
async globalChatter()
```

Macro code:

```js
game.voiceActorChatter.globalChatter()
```

macro code if you want to share the sounds with all the players

```js
game.voiceActorChatter.globalChatter({toAll: true})
```

macro code if you want to check the language withpolyglot given a explicit language

```js
game.voiceActorChatter.globalChatter({langs: ["gnomish", "gith"]})
```

## Global Chatter Every Interval

Every interval as measured in milliseconds, executes `globalChatter()`.

```js
randomGlobalChatterEvery(milliseconds)
```

Macro code:

```js
game.voiceActorChatter.randomGlobalChatterEvery(milliseconds)
```

macro code if you want to share the sounds with all the players

```js
game.voiceActorChatter.randomGlobalChatterEvery(milliseconds, {toAll: true})
```

macro code if you want to check the language withpolyglot given a explicit language

```js
game.voiceActorChatter.randomGlobalChatterEvery(milliseconds, {langs: ["gnomish", "gith"]});
```

## Disable Global Chatter

Clears out the timer on `randomGlobalChatterEvery`

```js
turnOffGlobalTimerChatter()
```

Macro code:

```js
game.voiceActorChatter.turnOffGlobalTimerChatter()
```

## Token Chatter

Given a `Token`, tries to find a matching Chatter Table. If none, exits. If one or more, randomly picks one and displays a rolled result from it as a ChatBubble.

```js
async tokenChatter(token)
```

Macro code:

```js
var token = game.scenes.filter(x => x.active)[0].data.tokens.filter(x => x.name == "Human Thug C")[0];
game.voiceActorChatter.tokenChatter(token);
```

macro code if you want to share the sounds with all the players

```js
var token = game.scenes.filter(x => x.active)[0].data.tokens.filter(x => x.name == "Human Thug C")[0];
game.voiceActorChatter.tokenChatter(token, {toAll: true});
```

macro code if you want to check the language withpolyglot given a explicit language

```js
var token = game.scenes.filter(x => x.active)[0].data.tokens.filter(x => x.name == "Human Thug C")[0];
game.voiceActorChatter.tokenChatter(token, {langs: ["gnomish", "gith"]});
```

## Selected Chatter

Grabs the currently selected Tokens and tries to find matching Chatter Tables. If none, exits. If one or more, randomly picks one and an elibable Token and displays a rolled result from the Table as a ChatBubble.

```js
  async selectedChatter()
```

Macro code:

```js
game.voiceActorChatter.selectedChatter()
```

macro code if you want to share the sounds with all the players

```js
game.voiceActorChatter.selectedChatter(token, {toAll: true});
```

macro code if you want to check the language withpolyglot given a explicit language

```js
game.voiceActorChatter.selectedChatter(token, {langs: ["gnomish", "gith"]});
```

# Build

## Install all packages

```bash
npm install
```
## npm build scripts

### build

will build the code and copy all necessary assets into the dist folder and make a symlink to install the result into your foundry data; create a
`foundryconfig.json` file with your Foundry Data path.

```json
{
  "dataPath": "~/.local/share/FoundryVTT/"
}
```

`build` will build and set up a symlink between `dist` and your `dataPath`.

```bash
npm run-script build
```

### NOTE:


CHANGELOG.md
# 0.2.7

- Better automatization on the creation of the rolltable with polyglot
- Add Explicit settings for dnd5e system with check the languages on the token

# 0.2.6

- Added automatic creation of the roll table for the polyglot integration
- Some bug fix and performance

# 0.2.5

- Some rename, bug fix and performance
- Update readme

# 0.2.4

- Integration with polyglot module

# 0.2.0 - 

- Convert project to typescript
- Integration of ecma scirpt module
- Add mutliclips management
- Add integration with rolltable for random sounds (inspired from npc chatter)
- Add custom directory as module settings
- Add integration with rolltable

# 0.1.2 - 2021/08/09

* Fixed broken audio files, republished

# 0.1.0 - 2021/05/22

* Fixed playback on 0.8+

# 0.0.9 - 2021/02/24

* Added check for insecure sites. Updated readme with workarounds.

# 0.0.8 - 2021/01/14

* Added support to record/playback in journal entries

# 0.0.7 - 2021/01/13

* Added option to allow players to record to their owned actors
* Added option to allow players to playback from limited/observer actors

# 0.0.6 - 2021/01/11

* Added forge support
* Updated max clip time to 30 seconds
* Fixed some styling issues
* Reduced minimum core version
* Switched to Interface/SoundBoard volume slider from Ambient

# 0.0.4 - 2021/01/10

* Added broadcast audio via shift-click

# 0.0.3 - 2021/01/10

* Disallowed non-gm interaction
* Added more readme details

# 0.0.2 - 2021/01/10

* Fixed localization

# 0.0.1 - 2021/01/10

* Initial Release


src/module/voiceactor.ts
/**
 * VoiceActor by Blitz
 */

import { TokenData } from '@league-of-foundry-developers/foundry-vtt-types/src/foundry/common/data/module.mjs';
import { BaseTableResult } from '@league-of-foundry-developers/foundry-vtt-types/src/foundry/common/documents.mjs';
import { i18n, warn } from '../main';
import { getGame } from './helpers';
import { polyglotIsActive } from './Hooks';
import { VOICE_ACTOR_CHATTER_MODULE_NAME } from './settings';
import { VoiceActorChatter } from './voiceactorchatter';

export class VoiceActor {
  static moduleName = VOICE_ACTOR_CHATTER_MODULE_NAME;

  static getClip = async (
    data: Token,
    customDirectory: string,
    isJournal: boolean,
  ): Promise<{ file: string; name: string }> => {
    if (!customDirectory) {
      customDirectory = <string>getGame().settings.get(VOICE_ACTOR_CHATTER_MODULE_NAME, 'customDirectory') ?? '';
    }

    if (customDirectory) {
      const nameActorFolder = VoiceActor.getClipActorFolderName(data);

      // Create directories
      try {
        await FilePicker.createDirectory(
          //@ts-ignore
          VoiceActor.isForge() ? 'forgevtt' : 'data',
          `${customDirectory}`,
          {},
        );
      } catch (e) {
        // DO NOTHING
      }
      try {
        await FilePicker.createDirectory(
          //@ts-ignore
          VoiceActor.isForge() ? 'forgevtt' : 'data',
          `${customDirectory}/`,
          {},
        );
      } catch (e) {
        // DO NOTHING
      }
      try {
        await FilePicker.createDirectory(
          //@ts-ignore
          VoiceActor.isForge() ? 'forgevtt' : 'data',
          `${customDirectory}/Journal`,
          {},
        );
      } catch (e) {
        // DO NOTHING
      }
      try {
        await FilePicker.createDirectory(
          //@ts-ignore
          VoiceActor.isForge() ? 'forgevtt' : 'data',
          `${customDirectory}/${nameActorFolder}`,
          {},
        );
      } catch (e) {
        // DO NOTHING
      }

      try {
        await FilePicker.createDirectory(
          //@ts-ignore
          VoiceActor.isForge() ? 'forgevtt' : 'data',
          `${customDirectory}/Journal/${nameActorFolder}`,
          {},
        );
      } catch (e) {
        // DO NOTHING
      }

      // Get files
      const vaDir = await FilePicker.browse(
        //@ts-ignore
        VoiceActor.isForge() ? 'forgevtt' : 'data',
        `${customDirectory}${isJournal ? '/Journal' : ''}/${nameActorFolder}`,
      );
      // Check if file exists already
      // const fileName = VoiceActor.getClipActorFileName(data, isJournal);
      const fileName = `${data.actor?._id}-${data.actor?.name
        ?.replace(/[^a-z0-9]/gi, '_')
        .toLowerCase()}-${VoiceActor.generateUUID()}.wav`;
      // const clip = <string>VoiceActor.getFile(vaDir.files, fileName);
      // let clip = <string>vaDir.files.find((el) => el.includes(fileName));
      // if (!clip) {
      //   clip = vaDir.files[0];
      //   fileName = clip.substring(clip.lastIndexOf('/') + 1);
      // }
      const clip = vaDir.target + '/' + fileName;

      return {
        file: clip,
        name: fileName,
      };
    } else {
      return {
        file: '',
        name: '',
      };
    }
  };

  static getClipActorFolderName = function (data: Token) {
    const fileName = `${data.actor?._id}-${data.actor?.name?.replace(/[^a-z0-9]/gi, '_').toLowerCase()}`;
    return fileName;
  };

  // static getClipActorFileName = function (data: Token, isJournal: boolean) {
  //   // Check if file exists already
  //   // let fileName;
  //   // if (isJournal) {
  //   //   fileName = `${data.actor?._id}.wav`;
  //   // } else {
  //     // if (data.actorLink) {
  //     //   fileName = `${data.actor.id}.wav`;
  //     // } else {
  //     // fileName = `${data.actor?._id}-${data.actor?.name?.replace(/[^a-z0-9]/gi, '_').toLowerCase()}.wav`;
  //     // }
  //   // }
  //   const fileName = `${data.actor?._id}-${data.actor?.name?.replace(/[^a-z0-9]/gi, '_').toLowerCase()}-${VoiceActor.generateUUID()}.wav`;
  //   return fileName;
  // };

  static generateUUID = function () {
    // Public Domain/MIT
    let d = new Date().getTime(); //Timestamp
    let d2 = (typeof performance !== 'undefined' && performance.now && performance.now() * 1000) || 0; //Time in microseconds since page-load or 0 if unsupported
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
      let r = Math.random() * 16; //random number between 0 and 16
      if (d > 0) {
        //Use timestamp until depleted
        r = (d + r) % 16 | 0;
        d = Math.floor(d / 16);
      } else {
        //Use microseconds since page-load if supported
        r = (d2 + r) % 16 | 0;
        d2 = Math.floor(d2 / 16);
      }
      return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
    });
  };

  // static getFile = (filesArray: string[], filename: string) => {
  //   const file = filesArray.find((el) => el.includes(filename));
  //   return file || '';
  // };

  static isForge = () => {
    //@ts-ignore
    if (typeof ForgeVTT !== 'undefined') {
      //@ts-ignore
      return ForgeVTT.usingTheForge;
    } else {
      return false;
    }
  };

  static playClipRandomFromToken = async function (token: Token) {
    const voiceActorFolder = <Folder>(
      getGame().folders?.contents.filter(
        (x) => x.type == 'RollTable' && x.name?.toLowerCase() == token.name?.toLowerCase() + ' voice actor',
      )[0]
    );
    const tables = <RollTable[]>(
      getGame().tables?.contents.filter(
        (x) => x.name?.toLowerCase().endsWith('voice') || x.data.folder == voiceActorFolder._id,
      )
    );
    const eligableTables: RollTable[] = tables.filter((t: RollTable) =>
      token.name?.toLowerCase().includes(<string>t.name?.toLowerCase().replace(' voice', '').trim()),
    );
    const table: RollTable = eligableTables[0];

    const roll = await table.data.document.roll();
    const clip = roll.results[0].data.text;
    this.playClip(clip, true);
  };

  static playClip = async function (clip: string, toAllWithSocket: boolean) {
    // let playVolume = getGame().settings.get("core", "globalInterfaceVolume"); // TODO CUSTOMIZE WITH MODULE SETTINGS ???

    // Audio file to be played back
    let vaPlaybackFile;
    //@ts-ignore
    const hasHowler = typeof Howl != 'undefined';
    if (clip) {
      // Used for onend and onstop
      const onFinish = (id) => {
        // Prevent caching, in case the user overwrites the clip
        if (vaPlaybackFile) {
          if (hasHowler) {
            vaPlaybackFile.unload();
          }
          vaPlaybackFile = undefined;
        }
        // vaStates.playing = false;
        // title.find("#voiceactor-playback #voiceactor-playback-icon").removeClass('fa-stop').addClass('fa-play');
      };
      // Play file
      const payload = {
        src: clip,
        volume: getGame().settings.get('core', 'globalInterfaceVolume'), // TODO CUSTOMIZE WITH MODULE SETTINGS ???
        onend: onFinish,
        onstop: onFinish,
      };
      if (hasHowler) {
        //@ts-ignore
        vaPlaybackFile = new Howl(payload);
        vaPlaybackFile.play();
      } else {
        //@ts-ignore
        vaPlaybackFile = new Sound(payload.src);
        vaPlaybackFile.on('end', onFinish);
        vaPlaybackFile.on('stop', onFinish);
        await vaPlaybackFile.load();
        vaPlaybackFile.play({ volume: payload.volume });
      }
      // vaStates.playing = true;
      // title.find("#voiceactor-playback #voiceactor-playback-icon").removeClass('fa-play').addClass('fa-stop');
      if (toAllWithSocket) {
        getGame().socket?.emit('playAudio', payload);
        ui.notifications?.notify(i18n('foundryvtt-voice-actor-chatter.notif.broadcasted'));
      } else {
        //AudioHelper.play({src: fileClipPlayPath, volume: playVolume, autoplay: true, loop: false}, true);
      }
    } else {
      ui.notifications?.notify(i18n('foundryvtt-voice-actor-chatter.notif.no-clip-for-actor'));
    }
  };

  // static retrieveOrCreateRollTable = async function (tokenData: Token, fileNamePath: string, fileName: string):Promise<RollTable> {
  static retrieveOrCreateRollTable = async function (actor: Actor, actorRollTableName: string): Promise<RollTable> {
    const voiceActorFolder = await VoiceActor.retrieveOrCreateRollTableFolder(actor);
    let myTable: RollTable | undefined = getGame().tables?.contents.find(
      (table: RollTable) =>
        table.name?.toLowerCase() == actorRollTableName.toLowerCase() && table.folder == voiceActorFolder,
    );
    if (!myTable) {
      const formula = '1d20';
      const min = 1;
      const max = 20;
      // ui.notifications.notify(getGame().i18n.format("foundryvtt-voice-actor-chatter.notif.no-rolltable-for-actor", actorRollTableName));
      myTable = <RollTable>await RollTable.create({
        name: actorRollTableName,
        // description: actorRollTableName, // This appears on every roll in the chat!
        results: [],
        replacement: true,
        displayRoll: true,
        img: 'icons/svg/sound.svg', //"modules/EasyTable/easytable.png"
        folder: voiceActorFolder ? voiceActorFolder : '',
        formula: formula ? formula : min == 1 ? `1d${max}` : `1d${max - min + 1}+${min - 1}`,
        //sort: number,
        //permission: object,
        //flags: object
      });
      // //@ts-ignore
      // await myTable.normalize();
    }
    return myTable;
  };

  static retrieveOrCreateRollTableFolder = async function (actor: Actor): Promise<Folder> {
    let baseFolder = <Folder>(
      getGame().folders?.contents.filter(
        (x: Folder) => x.type == 'RollTable' && x.name?.toLowerCase() == 'voice actor',
      )[0]
    );
    if (!baseFolder) {
      baseFolder = <Folder>await Folder.create({ name: 'Voice Actor', type: 'RollTable', parent: null });
    }
    //@ts-ignore
    let voiceActorFolder = baseFolder.children.filter(
      (x: Folder) =>
        x.type == 'RollTable' && x.name?.toLowerCase() == actor?.token?.name?.toLowerCase() + ' voice actor',
    )[0];
    if (!voiceActorFolder) {
      voiceActorFolder = await Folder.create({
        name: <string>actor?.name + ' voice actor',
        type: 'RollTable',
        parent: baseFolder?.id,
        sorting: 'm',
      });
    }
    if (!voiceActorFolder) {
      ui.notifications?.error(
        `The '${VOICE_ACTOR_CHATTER_MODULE_NAME}' module requires a folder with name (case insensitive) '${actor?.token?.name?.toLowerCase()} voice actor' on the rollTable sidebar.`,
      );
    }
    return voiceActorFolder;
  };

  static async _onCreateResult(rollTable: RollTable, fileNamePath: string, fileName: string) {
    // event.preventDefault();

    // Save any pending changes
    // await this._onSubmit(event);

    // Get existing results
    const results: any[] = Array.from(rollTable.data.document.results.values());
    const last = results[results.length - 1];

    // Get weight and range data
    const weight = last ? last.data.weight || 1 : 1;
    const totalWeight = <number>results.reduce((t: any, r: any) => t + r.data.weight, 0) || 1;
    const minRoll = results.length ? Math.min(...results.map((r: any) => r.data.range[0])) : 0;
    const maxRoll = results.length ? Math.max(...results.map((r: any) => r.data.range[1])) : 0;

    // Determine new starting range
    const spread = maxRoll - minRoll + 1;
    const perW = Math.round(spread / totalWeight);
    const range = [maxRoll + 1, maxRoll + Math.max(1, weight * perW)];

    // Create the new Result
    const resultData: any = {
      label: fileName,
      type: last ? last.type : CONST.TABLE_RESULT_TYPES.TEXT,
      collection: last ? last.collection : null,
      weight: weight,
      range: range,
      drawn: false,
      text: fileNamePath,
      img: 'icons/svg/sound.svg',
    };
    return rollTable.data.document?.createEmbeddedDocuments('TableResult', [resultData]);
  }

  // static getCollection(collection) {
  //   let validCollection = ['Actor', 'Scene', 'Macro', 'Playlist', 'JournalEntry', 'RollTable', 'Item']
  //   if (validCollection.includes(collection)) {
  //       return collection
  //   }
  //   return '';
  // }

  // static getResultId(collection, text) {
  //     let resultId = '';
  //     let img = 'icons/svg/d20-black.svg'
  //     if (collection == 'Text' || !collection) {
  //         return [resultId, img];
  //     }
  //     let entity;
  //     switch (collection) {
  //         case 'Actor':
  //                 entity = getGame().actors.getName(text);
  //                 resultId = entity?.id||''
  //                 img = entity?.img||img;
  //             break;
  //         case 'Scene':
  //                 entity = getGame().scenes.getName(text);
  //                 resultId = entity?.id||''
  //                 img = entity?.img||img;
  //             break;
  //         case 'Macro':
  //                 entity = getGame().macros.getName(text);
  //                 resultId = entity?.id||''
  //                 img = entity?.data?.img||img;
  //             break;
  //         case 'Playlist':
  //                 entity = getGame().playlists.getName(text);
  //                 resultId = entity?.id||''
  //                 // img = entity?.img||img;
  //             break;
  //         case 'JournalEntry':
  //                 entity = getGame().journal.getName(text);
  //                 resultId = entity?.id||''
  //                 img = entity?.data?.img||img;
  //             break;
  //         case 'RollTable':
  //                 entity = getGame().tables.getName(text);
  //                 resultId = entity?.id||''
  //                 img = entity?.data?.img||img;
  //             break;
  //         case 'Item':
  //                 entity = getGame().items.getName(text);
  //                 resultId = entity?.id||''
  //                 img = entity?.img||img;
  //             break;
  //         default:
  //             break;
  //     }
  //     return [resultId, img];
  // }
}

export const onRender = async (app, html, data) => {
  const disableHeaderSheetButtons =
    <string>getGame().settings.get(VOICE_ACTOR_CHATTER_MODULE_NAME, 'disableHeaderSheetButtons') ?? false;
  if (disableHeaderSheetButtons) {
    return;
  }

  const customDirectory = <string>getGame().settings.get(VOICE_ACTOR_CHATTER_MODULE_NAME, 'customDirectory') ?? '';

  // Get window-title from html so we can prepend our buttons
  const title = html.find('.window-title');

  // Store recording and playback states
  const vaStates = {
    recording: false,
    playing: false,
  };

  // Audio file to be played back
  let vaPlaybackFile;

  // MediaRecorder
  let vaRecorder;

  // timeout to sop vaRecorder after 10 seconds if not stopped manually
  let vaRecorderTimeout;

  let isJournal = false;
  if (data.options.classes.indexOf('journal-sheet') > -1) {
    isJournal = true;
  }

  let buttons = ``;

  if (
    getGame().user?.isGM ||
    (data.owner &&
      getGame().settings?.get(VOICE_ACTOR_CHATTER_MODULE_NAME, 'playersRecordOwned') &&
      getGame().user?.hasPermission('FILES_UPLOAD'))
  ) {
    buttons += `<button id="voiceactor-record" class="voiceactor-button" title="${i18n(
      'foundryvtt-voice-actor-chatter.ui.button-tooltip-record',
    )}">
        <i id="voiceactor-record-icon" style="color: white" class="fas fa-microphone"></i>
        </button>`;
  }

  if (
    getGame().user?.isGM ||
    data.owner ||
    getGame().settings.get(VOICE_ACTOR_CHATTER_MODULE_NAME, 'playersPlaybackLimited')
  ) {
    buttons += `<button id="voiceactor-playback" class="voiceactor-button" title="${i18n(
      'foundryvtt-voice-actor-chatter.ui.button-tooltip-playback',
    )}">
        <i id="voiceactor-playback-icon" style="color: white" class="fas fa-play"></i>
        </button>`;
  }

  // Add buttons
  title.prepend(buttons);

  const obj = await VoiceActor.getClip(data, customDirectory, isJournal);
  if (obj.file == '' || obj.name == '') {
    ui.notifications?.notify(i18n('foundryvtt-voice-actor-chatter.notif.no-clip-for-actor'));
    return;
  }
  const clip = obj.file;
  const fileName = obj.name;

  if (clip) {
    // Change button color if this actor has a clip already
    title.find('#voiceactor-record #voiceactor-record-icon').css('color', 'lightgreen');
  }

  title.find('#voiceactor-record').click(async (ev) => {
    if (vaStates.recording) {
      // Stop recording if button is pressed while recording active
      vaRecorder.stop();
      return;
    }

    if (clip) {
      if (!ev.shiftKey) {
        // Notify user if record is clicked but clip exists. Bypass if SHIFT is held when clicking.
        ui.notifications?.notify(i18n('foundryvtt-voice-actor-chatter.notif.clip-exists'));
        return;
      } else {
        if (VoiceActor.isForge()) {
          ui.notifications?.notify(i18n('foundryvtt-voice-actor-chatter.notif.forge-cache'));
        }
      }
    }

    // const fileName = VoiceActor.getClipActorFileName(data, isJournal);

    if (!navigator.mediaDevices) {
      ui.notifications?.error(i18n('foundryvtt-voice-actor-chatter.notif.no-media-devices'));
    }
    // Record clip
    navigator.mediaDevices
      .getUserMedia({
        audio: true,
      })
      .then(async (stream) => {
        vaStates.recording = true;
        title.find('#voiceactor-record #voiceactor-record-icon').removeClass('fa-microphone').addClass('fa-stop');
        title.find('#voiceactor-record #voiceactor-record-icon').css('color', 'red');
        const chunks: any[] = [];
        vaRecorder = new MediaRecorder(stream);
        vaRecorder.ondataavailable = async (e: any) => {
          chunks.push(e.data);
          if (vaRecorder.state == 'inactive') {
            const blob = new Blob(chunks, {
              type: 'audio/wav',
            });
            const file = new File([blob], fileName, {
              type: 'audio/wav',
            });

            const nameActorFolder = VoiceActor.getClipActorFolderName(data);
            const dirName = `${customDirectory}${isJournal ? '/Journal' : ''}/${nameActorFolder}`;
            //@ts-ignore
            await FilePicker.upload(VoiceActor.isForge() ? 'forgevtt' : 'data', dirName, file);
            vaStates.recording = false;

            // Only really works with Firefox, chrome has some weird caching, requiring the user to wait about 30 seconds
            //@ts-ignore
            getGame().audio?.buffers.delete(`${dirName[0] == '/' ? dirName.substr(1) : dirName}/${fileName}`);

            title.find('#voiceactor-record #voiceactor-record-icon').removeClass('fa-stop').addClass('fa-microphone');
            title.find('#voiceactor-record #voiceactor-record-icon').css('color', 'lightgreen');
            vaRecorder = null; //delete vaRecorder;
            clearTimeout(vaRecorderTimeout);
            vaRecorderTimeout = null; //delete vaRecorderTimeout;
            // stream = null; //delete stream;
            const fileNamePath = `${dirName[0] == '/' ? dirName.substr(1) : dirName}/${fileName}`;
            const actorRollTableName = data.actor?.name + ' voice';
            const theTable = await VoiceActor.retrieveOrCreateRollTable(data.actor, actorRollTableName);
            if (polyglotIsActive) {
              await VoiceActorChatter._createAutomaticPolyglotRolltable([data]);
            }
            VoiceActor._onCreateResult(theTable, fileNamePath, fileName);
          }
        };
        vaRecorder.start();
        // Stop recording after 30 seconds. Timeout is cancelled if user stops manually
        vaRecorderTimeout = setTimeout(() => {
          vaRecorder.stop();
        }, 30000);
      })
      .catch((e) => {
        console.log(e);
      });
  });

  title.find('#voiceactor-playback').click(async (ev: any) => {
    if (vaStates.playing) {
      // Stop playback if pressed while playing
      vaPlaybackFile.stop();
      vaStates.playing = false;
      return;
    }

    //@ts-ignore
    const hasHowler = typeof Howl != 'undefined';
    if (clip) {
      // Used for onend and onstop
      const onFinish = (id) => {
        // Prevent caching, in case the user overwrites the clip
        if (vaPlaybackFile) {
          if (hasHowler) {
            vaPlaybackFile.unload();
          }
          vaPlaybackFile = undefined;
        }
        vaStates.playing = false;
        title.find('#voiceactor-playback #voiceactor-playback-icon').removeClass('fa-stop').addClass('fa-play');
      };
      // Play file
      const payload = {
        src: clip,
        volume: getGame().settings.get('core', 'globalInterfaceVolume'), // TODO CUSTOMIZE WITH MODULE SETTINGS ???
        onend: onFinish,
        onstop: onFinish,
      };
      if (hasHowler) {
        //@ts-ignore
        vaPlaybackFile = new Howl(payload);
        vaPlaybackFile.play();
      } else {
        //@ts-ignore
        vaPlaybackFile = new Sound(payload.src);
        vaPlaybackFile.on('end', onFinish);
        vaPlaybackFile.on('stop', onFinish);
        await vaPlaybackFile.load();
        vaPlaybackFile.play({ volume: payload.volume });
      }
      vaStates.playing = true;
      title.find('#voiceactor-playback #voiceactor-playback-icon').removeClass('fa-play').addClass('fa-stop');
      if (ev.shiftKey) {
        getGame().socket?.emit('playAudio', payload);
        ui.notifications?.notify(i18n('foundryvtt-voice-actor-chatter.notif.broadcasted'));
      }
    } else {
      ui.notifications?.notify(i18n('foundryvtt-voice-actor-chatter.notif.no-clip-for-actor'));
    }
  });
};
